File: Python/Two Sum.py
"""
Given an array of integers, find two numbers such that they add up to a specific target number.
The function twoSum should return indices of the two numbers such that they add up to the target, where index1 must be less than index2. Please note that your returned answers (both index1 and index2) are not zero-based.
You may assume that each input would have exactly one solution.
Input: numbers={2, 7, 11, 15}, target=9
Output: index1=1, index2=2
"""

class Solution:
    def twoSum(self, nums, target):
        lookup = {}
        for i, num in enumerate(nums):
            if target - num in lookup:
                return (lookup[target - num] + 1, i + 1)
            lookup[num] = i


class Solution:
    # @return a tuple, (index1, index2)
    def twoSum(self, num, target):
        d = {}                          # This is used because we need to sort the array
        for i, n in enumerate(num):
            d.setdefault(n, []).append(i+1)
        num = sorted(num)
        l = 0
        r = len(num) - 1
        while l < r:
            if num[l] + num[r]  == target:
                if num[l] == num[r]:
                    return (d[num[l]][0], d[num[r]][1])
                else:
                    return sorted((d[num[l]][0], d[num[r]][0]))
            elif num[l] + num[r] < target:
                l += 1
            else:
                r -= 1

File: Java/Two Sum.java
/*
Given an array of integers, find two numbers such that they add up to a specific target number.

The function twoSum should return indices of the two numbers such that they add up to the target, where index1 must be less than index2. Please note that your returned answers (both index1 and index2) are not zero-based.

You may assume that each input would have exactly one solution.

Input: numbers={2, 7, 11, 15}, target=9
Output: index1=1, index2=2
*/

// Hash table: O(n) runtime, O(n) space
public class Solution {
    public int[] twoSum(int[] numbers, int target) {
        HashMap <Integer,Integer>table = new HashMap<Integer,Integer>();
        for(int i = 0;i<numbers.length;i++)
        {
            if(!table.containsKey(numbers[i]))
            {
                table.put(target - numbers[i],i);
            }
            else
            {
                int idx = table.get(numbers[i]);
                return new int[]{Math.min(idx,i)+1,Math.max(idx,i)+1};
            }
        }
        throw new IllegalArgumentException("no 2sum solution");
    }
}